# Aggrégation du fichier consommations

Ici se trouve la démarche et l'analyse exploratoire préalable réalisée pour l'agrégation du fichier consommations

In [2]:
import pandas as pd

df_consommations = pd.read_csv("s3://projet-bdc-data/alptis/training/apprentissage_consommations_2023_11.csv", sep=";")

In [3]:
len(df_consommations)

917832

In [4]:
df_consommations.columns

Index(['client_code', 'annee_mois_paiement', 'nb_decomptes',
       'nb_decomptes_hospitalisation', 'nb_decomptes_soins_medicaux',
       'nb_decomptes_pharmacie', 'nb_decomptes_optique',
       'nb_decomptes_dentaire', 'nb_decomptes_divers',
       'nb_jours_forfait_journalier', 'nb_jours_chambre_particuliere',
       'nb_protheses_dentaires', 'remb_alptis', 'remb_alptis_hospitalisation',
       'remb_alptis_soins_medicaux', 'remb_alptis_pharmacie',
       'remb_alptis_optique', 'remb_alptis_dentaire', 'remb_alptis_divers',
       'remb_ro', 'remb_ro_hospitalisation', 'remb_ro_soins_medicaux',
       'remb_ro_pharmacie', 'remb_ro_optique', 'remb_ro_dentaire',
       'remb_ro_divers', 'frais_reels', 'frais_reels_hospitalisation',
       'frais_reels_soins_medicaux', 'frais_reels_pharmacie',
       'frais_reels_optique', 'frais_reels_dentaire', 'frais_reels_divers',
       'reste_a_charge', 'reste_a_charge_hospitalisation',
       'reste_a_charge_soins_medicaux', 'reste_a_charge_pharmac

In [5]:
df_consommations.describe()

,client_code,nb_decomptes,nb_decomptes_hospitalisation,nb_decomptes_soins_medicaux,nb_decomptes_pharmacie,nb_decomptes_optique,nb_decomptes_dentaire,nb_decomptes_divers,nb_jours_forfait_journalier,nb_jours_chambre_particuliere,...,frais_reels_optique,frais_reels_dentaire,frais_reels_divers,reste_a_charge,reste_a_charge_hospitalisation,reste_a_charge_soins_medicaux,reste_a_charge_pharmacie,reste_a_charge_optique,reste_a_charge_dentaire,reste_a_charge_divers
count,9.178320e+05,917832.000000,917832.000000,917832.000000,917832.000000,917832.000000,917832.000000,917832.000000,917832.000000,917832.000000,...,917832.000000,917832.000000,917832.000000,9.178320e+05,917832.000000,9.178320e+05,917832.000000,917832.000000,917832.000000,917832.000000
mean,1.259118e+07,3.436187,0.071011,1.415524,1.485534,0.052905,0.145779,0.265433,0.326331,0.098138,...,19.266977,42.244699,24.012964,3.973516e+01,3.894211,1.349219e+01,0.139245,6.205941,11.627142,4.376426
std,3.517230e+06,2.913165,0.326381,1.812704,1.430266,0.253541,0.433252,0.666661,3.065847,1.402597,...,106.720612,272.057037,190.193576,7.408956e+03,93.739678,7.406559e+03,3.583178,49.228930,130.479795,85.007355
min,2.173600e+04,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000e+00,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000
25%,1.109135e+07,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000e+00,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000
50%,1.394454e+07,3.000000,0.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000e+00,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000
75%,1.520184e+07,5.000000,0.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,2.560000e+00,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000
max,1.619090e+07,77.000000,16.000000,50.000000,75.000000,6.000000,8.000000,51.000000,246.000000,151.000000,...,6516.750000,32020.000000,37499.870000,7.092276e+06,34752.510000,7.092276e+06,2053.130000,6246.750000,18014.000000,25685.840000


Après analyse rapide du fichier:

- 917832 lignes  
- 40 variables : dont identifiant client_code

**6 catégories de soins:** 
- Hospitalisation 
- soins_medicaux 
- pharmacie 
- optique 
- dentaire 
- divers --> voir avec Alptis à quoi ça correspond 

A chaque fois, pour chaque catégorie on a des informations sur:  
- ***nbre_decompte*** : le nbre de remboursement réalisé sur des soins : plus précisément le nombre de fois où la Sécurité social a envoyé (ou transmis à la mutuelle) un relevé de remboursement. 
- ***remb_alptis*** : montant remboursement payé par Alptis 
- ***remb_ro*** : remeboursement par le Régime Obligataire = sécurité sociale 
- ***frais_rééls*** : montant frais total engagés (somme remboursement alptis + RO ) 
- ***reste_ a_charge*** : ce que le client doit payer après remboursement RO et Alptis 

On a les montants et le nombre de décompte pour chaque catégorie de soin, mais également le nombre de décompte total et le montant des soins totaux, toute catégories de soin confondues. 

Remarques importantes :  
-  **Remb_alptis+ remb_ro + reste_a_charge= frais_reels**
- **Nb_decomptes = Nb_decomptes_pharmacie + Nb_decomptes_optique + Nb_decomptes_soins_medicaux + Nb_decomptes_hospitalisation + Nb_decomptes_divers**

In [6]:
df_consommations.dtypes

client_code                         int64
annee_mois_paiement                object
nb_decomptes                      float64
nb_decomptes_hospitalisation      float64
nb_decomptes_soins_medicaux       float64
nb_decomptes_pharmacie            float64
nb_decomptes_optique              float64
nb_decomptes_dentaire             float64
nb_decomptes_divers               float64
nb_jours_forfait_journalier       float64
nb_jours_chambre_particuliere     float64
nb_protheses_dentaires            float64
remb_alptis                       float64
remb_alptis_hospitalisation       float64
remb_alptis_soins_medicaux        float64
remb_alptis_pharmacie             float64
remb_alptis_optique               float64
remb_alptis_dentaire              float64
remb_alptis_divers                float64
remb_ro                           float64
remb_ro_hospitalisation           float64
remb_ro_soins_medicaux            float64
remb_ro_pharmacie                 float64
remb_ro_optique                   

In [7]:
#transformation de annee_mois_paiement en colonne str
df_consommations["annee_mois_paiement"] = (
    df_consommations["annee_mois_paiement"]
    .astype("string")
    .str.strip()
)

#extraction de toutes les colonnes sauf code_client et annee_mois_paiement
cols_num = df_consommations.select_dtypes(include=["float64"]).columns.tolist()

#Aggrégation en faisant la somme sur toutes les lignes, sur chaque colonne selon la valeur 
#code_client pour les colonnes de cols_num, et en faisant une liste de str pour annee_mois_paiement
df_consommations_agg = df_consommations.groupby("client_code").agg(
    {
        **{col: "sum" for col in cols_num},              # somme robuste aux NaN
        "annee_mois_paiement": lambda x: sorted([v for v in x.dropna().unique()])  # liste sans NaN
    }
).reset_index()

In [8]:
#Vérification de la bonne aggrégation
df_agg= df_consommations_agg
print((df_agg["nb_decomptes"] == df_agg["nb_decomptes_dentaire"] + df_agg["nb_decomptes_pharmacie"] + df_agg["nb_decomptes_soins_medicaux"] + df_agg["nb_decomptes_optique"] + + df_agg["nb_decomptes_divers"] + df_agg["nb_decomptes_hospitalisation"]).all())
print(max([c for c in df_agg["client_code"].value_counts()])) #On a bien 1 code client par ligne

True
1


In [9]:
print(df_agg)
df_agg.columns

        client_code  nb_decomptes  nb_decomptes_hospitalisation  \
0             21736          23.0                           0.0   
1             22037          25.0                           0.0   
2             22331          84.0                           2.0   
3             22604          27.0                           0.0   
4             22702          65.0                           6.0   
...             ...           ...                           ...   
120902     16184393           3.0                           0.0   
120903     16184764           1.0                           0.0   
120904     16188411           1.0                           0.0   
120905     16190238           1.0                           0.0   
120906     16190896           1.0                           0.0   

        nb_decomptes_soins_medicaux  nb_decomptes_pharmacie  \
0                               2.0                    20.0   
1                              11.0                    14.0   
2     

Index(['client_code', 'nb_decomptes', 'nb_decomptes_hospitalisation',
       'nb_decomptes_soins_medicaux', 'nb_decomptes_pharmacie',
       'nb_decomptes_optique', 'nb_decomptes_dentaire', 'nb_decomptes_divers',
       'nb_jours_forfait_journalier', 'nb_jours_chambre_particuliere',
       'nb_protheses_dentaires', 'remb_alptis', 'remb_alptis_hospitalisation',
       'remb_alptis_soins_medicaux', 'remb_alptis_pharmacie',
       'remb_alptis_optique', 'remb_alptis_dentaire', 'remb_alptis_divers',
       'remb_ro', 'remb_ro_hospitalisation', 'remb_ro_soins_medicaux',
       'remb_ro_pharmacie', 'remb_ro_optique', 'remb_ro_dentaire',
       'remb_ro_divers', 'frais_reels', 'frais_reels_hospitalisation',
       'frais_reels_soins_medicaux', 'frais_reels_pharmacie',
       'frais_reels_optique', 'frais_reels_dentaire', 'frais_reels_divers',
       'reste_a_charge', 'reste_a_charge_hospitalisation',
       'reste_a_charge_soins_medicaux', 'reste_a_charge_pharmacie',
       'reste_a_charge_op

In [10]:
#Remarque:  nb_decomptes =/= nbre de annee_mois_paiement
df_consommations_agg[["client_code", "nb_decomptes", "frais_reels","annee_mois_paiement"]]

,client_code,nb_decomptes,frais_reels,annee_mois_paiement
0,21736,23.0,1181.26,"[2022-12, 2023-02, 2023-03, 2023-04, 2023-05, ..."
1,22037,25.0,1642.47,"[2022-12, 2023-01, 2023-02, 2023-03, 2023-04, ..."
2,22331,84.0,4509.60,"[2022-12, 2023-01, 2023-02, 2023-03, 2023-04, ..."
3,22604,27.0,1550.61,"[2022-12, 2023-01, 2023-03, 2023-04, 2023-05, ..."
4,22702,65.0,5959.13,"[2022-12, 2023-01, 2023-02, 2023-03, 2023-04, ..."
...,...,...,...,...
120902,16184393,3.0,96.95,[2023-11]
120903,16184764,1.0,26.50,[2023-11]
120904,16188411,1.0,26.50,[2023-11]
120905,16190238,1.0,120.00,[2023-11]


Nous avons crée le DataFrame *df_consommations_agg*, via la démarche suivante: 
- Pour toutes les colonnes sauf annee_mois_paiement et client_code: on fait la somme sur toutes les lignes d'un même code_client
- Pour annee_mois_paiement: pour chaque ligne code_client, avoir une liste des annee_mois_paiement des remboursements réalisés par alptis. 
Ainsi on conserve l'information concernant la date et la fréquence des remboursements. --> Utile pour analyser si client consulte énormément ou pas et si pic de consultation à un moment dans l'année.

**Remarque:** Le nombre de dates de paiement enregistrées dans l'année pour un code_client particulier (*annee_mois_paiement*) ne correspond pas au nombre de décomptes totaux dans l'année pour ce même code client (*nb_decomptes*). En effet, un client peut recevoir divers remboursements, provenant de soins divers (de la même catégorie ou non) pendant le même mois.


**Problème**:
Cette aggégration nous fait perdre l'information du type de soin et du montant payé (par Alptis, le RO et le client) associé à chaque date de remboursement.
Tentons de la conserver en créant une nouvelle colonne ***info_par_annee_paiement*** dans le DataFrame, avec chaque ligne au format suivant: 

***{anne_mois_paiement: soin : {rb_alptis, rb_ro, reste_a_charge, frais_reels}}*** 

avec soin dans [hospitalisation, dentaire, hospitalisation, optique, soins_medicaux, divers]

In [11]:
#Création d'un dictionnaire mapping_soin pour nous aider dans l'extraction des informations
soins = ["hospitalisation", "soins_medicaux", "pharmacie", "optique", "dentaire", "divers"]
mapping_soins = {
    soin: {
        "frais_reels": f"frais_reels_{soin}",
        "reste_a_charge": f"reste_a_charge_{soin}",
        "remb_alptis": f"remb_alptis_{soin}",
        "remb_ro": f"remb_ro_{soin}",
    }
    for soin in soins
}

#Création d'une fonction qui extrait pour chaque date, le soin associé et les montants associés
def build_info_dict(df_client):
    info = {}

    for _, row in df_client.iterrows():
        date = row["annee_mois_paiement"]

        if date not in info:
            info[date] = {}

        for soin, cols in mapping_soins.items():

            # Récupérer les 4 montants
            fr = row[cols["frais_reels"]]
            rac = row[cols["reste_a_charge"]]
            alpt = row[cols["remb_alptis"]]
            ro = row[cols["remb_ro"]]

            # Vérifier si le soin est présent (= au moins une valeur non nulle)
            if (fr != 0) or (rac != 0) or (alpt != 0) or (ro != 0):
                info[date][soin] = {
                    "frais_reels": fr,
                    "reste_a_charge": rac,
                    "remb_alptis": alpt,
                    "remb_ro": ro,
                }

        # Si aucun soin présent → supprimer la date vide
        if info[date] == {}:
            del info[date]

    return info

In [12]:
df_agg["consommation_info_par_annee_mois_paiement"] = (
    df_consommations.groupby("client_code")
    .apply(build_info_dict)
    .reset_index(drop=True)
)
df_consommations_agg=df_agg

/tmp/ipykernel_385953/244209272.py:3: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(build_info_dict)


In [13]:
#Test sur un code client (22037)
print(df_consommations[df_consommations["client_code"]==22037])

        client_code annee_mois_paiement  nb_decomptes  \
1             22037             2022-12           2.0   
61885         22037             2023-01           2.0   
127322        22037             2023-02           1.0   
193335        22037             2023-03           3.0   
268362        22037             2023-04           1.0   
341929        22037             2023-05           1.0   
416692        22037             2023-06           2.0   
498541        22037             2023-07           2.0   
578666        22037             2023-08           3.0   
655872        22037             2023-09           1.0   
740837        22037             2023-10           3.0   
826319        22037             2023-11           4.0   

        nb_decomptes_hospitalisation  nb_decomptes_soins_medicaux  \
1                                0.0                          1.0   
61885                            0.0                          1.0   
127322                           0.0               

In [14]:
#ligne associée dans le dictionnaire aggrégé
print(df_agg[df_agg["client_code"]==22037]["consommation_info_par_annee_mois_paiement"]) #1 seule ligne qui aggrège les infos sur les soins et les montants associés
print(df_agg[df_agg["client_code"]==22037]["consommation_info_par_annee_mois_paiement"][1]["2022-12"]) #valeur pour une date particulière

1    {'2022-12': {'soins_medicaux': {'frais_reels':...
Name: consommation_info_par_annee_mois_paiement, dtype: object
{'soins_medicaux': {'frais_reels': 54.84, 'reste_a_charge': 0.0, 'remb_alptis': 21.93, 'remb_ro': 32.91}, 'pharmacie': {'frais_reels': 11.96, 'reste_a_charge': 0.0, 'remb_alptis': 7.57, 'remb_ro': 4.39}}
